gcalbot/gcalbot/invite.go

```diff
@@ -149,3 +153,94 @@ func (h *Handler) updateEventResponseStatus(invite *Invite, account *Account, re
 
 	return nil
 }
+
+func (h *Handler) syncAllInvites(account *Account, srv *calendar.Service, channelID, calendarID string) {
+	syncStart := time.Now()
+
+	var nextSyncToken string
+	var events []*calendar.Event
+	err := srv.Events.List(calendarID).
+		Pages(context.Background(), func(page *calendar.Events) error {
+			if page.NextPageToken == "" {
+				// set the sync token when the page token is empty
+				nextSyncToken = page.NextSyncToken
+			}
+			events = append(events, page.Items...)
+			return nil
+		})
+	if err != nil {
+		h.Errorf("error syncing all invites: %s", err)
+		return
+	}
+
+	for _, event := range events {
+		status := EventStatus(event.Status)
+
+		// if the event is cancelled or there aren't any attendees (ie. the user created the event), skip - it's not an invite
+		if status == EventStatusCancelled || event.Attendees == nil {
+			continue
+		}
+
+		// if the event is recurring, only deal with the underlying recurring event
+		if event.RecurringEventId != "" && event.RecurringEventId != event.Id {
+			continue
+		}
+
+		var end time.Time
+		if event.End == nil {
+			h.Errorf("empty dates in event")
+			continue
+		} else if event.End.DateTime != "" {
+			// this is a normal event
+			end, err = time.Parse(time.RFC3339, event.End.DateTime)
+			if err != nil {
+				h.Errorf("error parsing time: %s", err)
+				continue
+			}
+		} else if event.End.Date != "" {
+			// this is an all day event
+			end, err = time.Parse(AllDayDateFormat, event.End.Date)
+			if err != nil {
+				h.Errorf("error parsing time: %s", err)
+				continue
+			}
+			end = end.Add(-24 * time.Hour) // the google API sets the end day to the day after, so compensate by one day
+		} else {
+			h.Errorf("invalid end date: %+v", event.End)
+			continue
+		}
+
+		if time.Now().After(end) {
+			// the event has already ended, don't send an invite
+			continue
+		}
+
+		for _, attendee := range event.Attendees {
+			responseStatus := ResponseStatus(attendee.ResponseStatus)
+			if attendee.Self && !attendee.Organizer && responseStatus == ResponseStatusNeedsAction {
+				err = h.db.InsertInvite(account, Invite{
+					CalendarID: calendarID,
+					EventID:    event.Id,
+				})
+				if mysqlErr, ok := err.(*mysql.MySQLError); ok {
```

Comment:
why not have the `InsertInvite` method have a `ON DUPLICATE KEY UPDATE...` statement?

Reply:
If the invite already exists, there will be a msgid attached to it - I wouldn't want to overwrite that with zero. I guess I could on duplicate key update nothing

Reply:
@joshblum do you think it would be better to "do nothing" on a duplicate key:
```sql
INSERT INTO invite
(keybase_username, account_nickname, calendar_id, event_id, message_id)
VALUES (?, ?, ?, ?, ?)
ON DUPLICATE KEY UPDATE
keybase_username=VALUES(keybase_username),
account_nickname=VALUES(account_nickname),
calendar_id=VALUES(calendar_id),
event_id=VALUES(event_id)
```

Reply:
you can just say `message_id=message_id` (not `VALUES(message_id)`) to noop

Reply:
I could just do that for all of them then? Effectively I'd want to ignore any changes (the key is everything that is not the msg_id)

Reply:
you don't need to do it for all of them, if you have a duplicate key you already know all of the other fields match, so you can just keep the old value of `message_id`

Reply:
Ok, I think I did what you were expecting but could you check?